docs/proposals/202005_query_logging.md

```diff
@@ -0,0 +1,257 @@
+
+The proposal for Query Logging has been divided into two major parts - Audit Logging and Adaptive Logging. Audit and Adaptive Logging would be implemented across the Store API level, and would also cover the HTTP APIs of querier and other components that use HTTP APIs.
+
+![](../img/thanos_log_limit.png)
+
+Each of the use case of the sub-part would be described and a possible implementation would be discussed -
+
+#### Using Middlewares -
+
+Since Middlewares are a good chunk of the implementation, it makes sense to discuss briefly about -
+
+* How we are going to use the middlewares in our logging.
+* How can users can configure it?
+* Discuss the different types of middlewares that we plan to use.
+
+**Usage of Middlewares** : As it has been suggested by the diagram above, we plan to put in interceptors/middlewares to intercept the connection for a server implementation and client implementation.
+
+The middlewares helps in intercepting the queries flowing before a client, and all the queries leaving the server.
+
+The middlewares provide a nice abstraction to log the queries because we have the following objective from the logging -
+
+* Log all the queries
+* Log some specific queries
+* Logging at a certain level(Debug / Info)
+* Have a same request-id for a given request
+
+All of the goals are easily possible using grpc middlewares, as the grpc-ecosystem has a nice set of middlewares that makes the above possible, so we can just plug and play our middlewares
+
+For HTTP, all these nice middlewares are not present, so we need to code up our implementation performing the same logic.
+
+**User specific configurations** : Currently we are going to provide the following user facing configurations for using the request logger - 
+* Use pre-defined adaptive logging policy.
+* The Adaptive Logging policy plans to include logging of only requests that return an error, or crosses a specific duration of time.
+* The level for high latency requests is `DEBUG`. Error, however would be logged at the `ERROR` level.
+* We would add in a flag to turn on/off request logging, and another flag to accept the log file path where we can log the details.
+* All the audit/adaptive logging can be logged into the respective files by passing the path of the files as a cli argument.
+* Request Logging would be turned off by default, and default logging would be in stdout. For logging all the requests, we would accept a flag for the enabling the same.
+* Currently the policy for request logging is hardcoded, may be extended as a user based policy later.
+
+**Types of Middlewares** : We are going to use the following middlewares for achieving the above objectives -
+
+* [`logging.WithDecider`](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/v2/interceptors/logging/options.go#L51) - This `grpc.Option` would help us in deciding the logic whether a given request should be logged or not. This should help in logging specific / all queries
```

Comment:
What about pre request logic? Can I control it?

Reply:
I need to think about that optionality :P